src/utils/mergeDeep.js

```diff
@@ -0,0 +1,30 @@
+const isObject = (obj) => obj && typeof obj === 'object';
+
+/**
+ * Performs a deep merge of objects and returns new object.
+ *
+ * @param {...object} objects
+ * @returns {object}
+ */
+export const mergeDeep = (...objects) => objects.reduce((prev, obj) => {
+  if (obj == null) {
+    return prev;
+  }
+
+  const newObject = { ...prev };
+
+  Object.keys(obj).forEach((key) => {
+    const pVal = prev[key];
+    const oVal = obj[key];
+
+    if (Array.isArray(pVal) && Array.isArray(oVal)) {
+      newObject[key] = pVal.concat(...oVal);
```

Comment:
Not sure whether we want to concat arrays or just replace them.  Props are typically plain values or objects so it can be omitted. What do you think @adamkudrna @mbohal @dacerondrej?

Reply:
I think we want replace rather the `concat`. 
Replace is safer, because otherwise there is no way to remove something from an array prop for a subtree of the app. I think this needs to be changed here.

In case we need more flexibility, we could in the future allow dynamic prop value definition. Maybe the value of a prop could be defined by a function accepting the `oldProp` as an argument.